com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.refactor.cs

```diff
@@ -0,0 +1,262 @@
+        /// <param name="timelineRenames">Timeline rename mapping</param>
+        void UpgradeRemaining(
+            Dictionary<int, List<ConversionLink>> conversionLinksPerScene,
+            Dictionary<string, string> timelineRenames)
+        {
+            for (var s = 0; s < m_SceneManager.SceneCount; ++s)
+            {
+                var scene = OpenScene(s);
+                var timelineManager = new TimelineManager(scene);
+                var upgradedObjects = new HashSet<GameObject>();
+                
+                UpgradePrefabInstances(upgradedObjects, conversionLinksPerScene[s], timelineManager, false);
+                
+                var rootObjects = scene.GetRootGameObjects();
+                var upgradables = GetUpgradables(rootObjects, m_ObjectUpgrader.RootUpgradeComponentTypes, true);
+                UpgradeNonPrefabs(upgradables, upgradedObjects, timelineManager);
+                
+#if CINEMACHINE_TIMELINE
+                var playableDirectors = TimelineManager.GetPlayableDirectors(scene);
+                UpdateAnimationReferences(playableDirectors);
+#endif
+                UpgradeObjectReferences(rootObjects);
+                
+                // Clean up all obsolete components
+                foreach (var go in rootObjects)
+                    m_ObjectUpgrader.DeleteObsoleteComponents(go);
+                
+#if CINEMACHINE_TIMELINE
+                // Restore timeline names
+                foreach (var director in playableDirectors)
+                {
+                    if (timelineRenames.ContainsKey(director.name)) // search based on guid name
+                    {
+                        director.name = timelineRenames[director.name]; // restore director name
+                    }
+                }
+#endif
+                EditorSceneManager.SaveScene(scene);
+            }
+
+            m_CurrentSceneOrPrefab = string.Empty;
+        }
+
+        /// <summary>
+        /// For each prefab asset
+        /// - Delete obsolete components
+        /// - Update manager camera caches
+        /// </summary>
+        void CleanupPrefabAssets()
+        {
+            for (var p = 0; p < m_PrefabManager.PrefabCount; ++p)
+            {
+                m_CurrentSceneOrPrefab = m_PrefabManager.GetPrefabAssetPath(p);
+                using var editingScope = new PrefabUtility.EditPrefabContentsScope(m_CurrentSceneOrPrefab);
+                var prefabContents = editingScope.prefabContentsRoot;
+                var components = new List<Component>();
+                foreach (var type in m_ObjectUpgrader.RootUpgradeComponentTypes)
+                    components.AddRange(prefabContents.GetComponentsInChildren(type, true).ToList());
+                
+                foreach (var c in components)
+                {
+                    if (c == null || c.gameObject == null)
+                        continue; // was a hidden rig
```

Comment:
this is not needed

Reply:
When a freelook is upgraded, the components will have null refs (that were pointing to the children).

Reply:
I don't think so, because you have just gotten them from `prefabContents.GetComponentsInChildren` which will not return null objects, and no upgrade has happened since then

Reply:
UpgradeObjectComponents called UpgradeFreelook which destroys the child rigs.

Reply:
When do you call `UpgradeObjectComponents`?

Reply:
Just under this `if (c == null || c.gameObject == null)`
```
foreach (var c in components)
{
    if (c == null || c.gameObject == null)
        continue; // was a hidden rig
    if (c.GetComponentInParent<CinemachineDoNotUpgrade>(true) != null)
        continue; // is a backup copy

    // Upgrade prefab and fix timeline references
    UpgradeObjectComponents(c.gameObject, timelineManager);
}
```

Reply:
Are we looking at the same thing?  This is what I see:
```                foreach (var c in components)
                {
                    if (c == null || c.gameObject == null)
                        continue; // was a hidden rig
                    if (c.GetComponentInParent<CinemachineDoNotUpgrade>(true) != null)
                        continue; // is a backup copy

                    m_ObjectUpgrader.DeleteObsoleteComponents(c.gameObject);
                }```

Reply:
Ohhh, no we were not looking at the same thing! 😵‍💫 Sorry! You are right, we don't need this check here! Removed!